// ngx-toast-notifications/esm5/core/toast-natifications.js
/**
 * @fileoverview added by tsickle
 * Generated from: core/toast-natifications.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Toaster } from '../toaster';
var ToastNotifications = /** @class */ (function () {
    function ToastNotifications(_toaster) {
        this._toaster = _toaster;
    }
    /**
     * @deprecated since version 1.0.0
     */
    /**
     * @deprecated since version 1.0.0
     * @param {?} toast
     * @return {?}
     */
    ToastNotifications.prototype.next = /**
     * @deprecated since version 1.0.0
     * @param {?} toast
     * @return {?}
     */
    function (toast) {
        /** @type {?} */
        var config = {
            text: toast.text,
            caption: toast.caption,
            type: toast.type,
            duration: toast.duration || toast.lifetime,
            component: null,
        };
        this._toaster.open(config);
    };
    ToastNotifications.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ToastNotifications.ctorParameters = function () { return [
        { type: Toaster }
    ]; };
    return ToastNotifications;
}());
export { ToastNotifications };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ToastNotifications.prototype._toaster;
}
