// ngx-toast-notifications/esm2015/core/toast-natifications.js
/**
 * @fileoverview added by tsickle
 * Generated from: core/toast-natifications.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Toaster } from '../toaster';
/**
 * @deprecated since version 1.0.0 use Toaster
 */
export class ToastNotifications {
    /**
     * @param {?} _toaster
     */
    constructor(_toaster) {
        this._toaster = _toaster;
    }
    /**
     * @deprecated since version 1.0.0
     * @param {?} toast
     * @return {?}
     */
    next(toast) {
        /** @type {?} */
        const config = {
            text: toast.text,
            caption: toast.caption,
            type: toast.type,
            duration: toast.duration || toast.lifetime,
            component: null,
        };
        this._toaster.open(config);
    }
}
ToastNotifications.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ToastNotifications.ctorParameters = () => [
    { type: Toaster }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ToastNotifications.prototype._toaster;
}
