// angular-bootstrap-data-table/esm2015/lib/components/pagination.component.js
import { Component, Inject, forwardRef } from '@angular/core';
import { DataTable } from './table.component';
import * as i0 from "@angular/core";
import * as i1 from "./table.component";
import * as i2 from "@angular/forms";
import * as i3 from "../utils/min";
const _c0 = function (a0, a1) { return [a0, a1]; };
export class DataTablePagination {
    constructor(dataTable) {
        this.dataTable = dataTable;
    }
    pageBack() {
        this.dataTable.offset -= Math.min(this.dataTable.limit, this.dataTable.offset);
    }
    pageForward() {
        this.dataTable.offset += this.dataTable.limit;
    }
    pageFirst() {
        this.dataTable.offset = 0;
    }
    pageLast() {
        this.dataTable.offset = (this.maxPage - 1) * this.dataTable.limit;
    }
    get maxPage() {
        return Math.ceil(this.dataTable.itemCount / this.dataTable.limit);
    }
    get limit() {
        return this.dataTable.limit;
    }
    set limit(value) {
        this.dataTable.limit = Number(value); // TODO better way to handle that value of number <input> is string?
    }
    get page() {
        return this.dataTable.page;
    }
    set page(value) {
        this.dataTable.page = Number(value);
    }
}
DataTablePagination.ɵfac = function DataTablePagination_Factory(t) { return new (t || DataTablePagination)(i0.ɵɵdirectiveInject(forwardRef(() => DataTable))); };
DataTablePagination.ɵcmp = i0.ɵɵdefineComponent({ type: DataTablePagination, selectors: [["data-table-pagination"]], decls: 33, vars: 18, consts: [[1, "pagination-box"], [1, "pagination-range"], [3, "textContent"], [1, "pagination-controllers"], [1, "pagination-limit"], [1, "input-group"], [1, "input-group-addon"], ["type", "number", "min", "1", "step", "1", 1, "form-control", 3, "ngModel", "blur", "keyup.enter", "keyup.esc"], ["limitInput", ""], [1, "pagination-pages"], [1, "btn", "btn-default", "pagination-firstpage", 3, "disabled", "click"], [1, "btn", "btn-default", "pagination-prevpage", 3, "disabled", "click"], [1, "pagination-page"], ["type", "number", "min", "1", "step", "1", 1, "form-control", 3, "max", "ngModel", "blur", "keyup.enter", "keyup.esc"], ["pageInput", ""], [1, "btn", "btn-default", "pagination-nextpage", 3, "disabled", "click"], [1, "btn", "btn-default", "pagination-lastpage", 3, "disabled", "click"]], template: function DataTablePagination_Template(rf, ctx) { if (rf & 1) {
        const _r2 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵelementStart(1, "div", 1);
        i0.ɵɵtext(2);
        i0.ɵɵelement(3, "span", 2);
        i0.ɵɵtext(4, " - ");
        i0.ɵɵelement(5, "span", 2);
        i0.ɵɵpipe(6, "min");
        i0.ɵɵtext(7, " / ");
        i0.ɵɵelement(8, "span", 2);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(9, "div", 3);
        i0.ɵɵelementStart(10, "div", 4);
        i0.ɵɵelementStart(11, "div", 5);
        i0.ɵɵelementStart(12, "span", 6);
        i0.ɵɵtext(13);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(14, "input", 7, 8);
        i0.ɵɵlistener("blur", function DataTablePagination_Template_input_blur_14_listener() { i0.ɵɵrestoreView(_r2); const _r0 = i0.ɵɵreference(15); return ctx.limit = _r0.value; })("keyup.enter", function DataTablePagination_Template_input_keyup_enter_14_listener() { i0.ɵɵrestoreView(_r2); const _r0 = i0.ɵɵreference(15); return ctx.limit = _r0.value; })("keyup.esc", function DataTablePagination_Template_input_keyup_esc_14_listener() { i0.ɵɵrestoreView(_r2); const _r0 = i0.ɵɵreference(15); return _r0.value = ctx.limit; });
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(16, "div", 9);
        i0.ɵɵelementStart(17, "button", 10);
        i0.ɵɵlistener("click", function DataTablePagination_Template_button_click_17_listener() { return ctx.pageFirst(); });
        i0.ɵɵtext(18, " \u00AB ");
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(19, "button", 11);
        i0.ɵɵlistener("click", function DataTablePagination_Template_button_click_19_listener() { return ctx.pageBack(); });
        i0.ɵɵtext(20, " \u2039 ");
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(21, "div", 12);
        i0.ɵɵelementStart(22, "div", 5);
        i0.ɵɵelementStart(23, "input", 13, 14);
        i0.ɵɵlistener("blur", function DataTablePagination_Template_input_blur_23_listener() { i0.ɵɵrestoreView(_r2); const _r1 = i0.ɵɵreference(24); return ctx.page = _r1.value; })("keyup.enter", function DataTablePagination_Template_input_keyup_enter_23_listener() { i0.ɵɵrestoreView(_r2); const _r1 = i0.ɵɵreference(24); return ctx.page = _r1.value; })("keyup.esc", function DataTablePagination_Template_input_keyup_esc_23_listener() { i0.ɵɵrestoreView(_r2); const _r1 = i0.ɵɵreference(24); return _r1.value = ctx.page; });
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(25, "div", 6);
        i0.ɵɵelementStart(26, "span");
        i0.ɵɵtext(27, "/");
        i0.ɵɵelementEnd();
        i0.ɵɵelement(28, "span", 2);
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(29, "button", 15);
        i0.ɵɵlistener("click", function DataTablePagination_Template_button_click_29_listener() { return ctx.pageForward(); });
        i0.ɵɵtext(30, " \u203A ");
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(31, "button", 16);
        i0.ɵɵlistener("click", function DataTablePagination_Template_button_click_31_listener() { return ctx.pageLast(); });
        i0.ɵɵtext(32, " \u00BB ");
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx.dataTable.translations.paginationRange, ": ");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("textContent", ctx.dataTable.offset + 1);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("textContent", i0.ɵɵpipeBind1(6, 13, i0.ɵɵpureFunction2(15, _c0, ctx.dataTable.offset + ctx.dataTable.limit, ctx.dataTable.itemCount)));
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("textContent", ctx.dataTable.itemCount);
        i0.ɵɵadvance(5);
        i0.ɵɵtextInterpolate1("", ctx.dataTable.translations.paginationLimit, ":");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngModel", ctx.limit);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("disabled", ctx.dataTable.offset <= 0);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("disabled", ctx.dataTable.offset <= 0);
        i0.ɵɵadvance(4);
        i0.ɵɵpropertyInterpolate("max", ctx.maxPage);
        i0.ɵɵproperty("ngModel", ctx.page);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("textContent", ctx.dataTable.lastPage);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("disabled", ctx.dataTable.offset + ctx.dataTable.limit >= ctx.dataTable.itemCount);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("disabled", ctx.dataTable.offset + ctx.dataTable.limit >= ctx.dataTable.itemCount);
    } }, directives: [i2.NumberValueAccessor, i2.DefaultValueAccessor, i2.NgControlStatus, i2.NgModel], pipes: [i3.MinPipe], styles: [".pagination-box[_ngcontent-%COMP%]{position:relative;margin-top:-10px}.pagination-range[_ngcontent-%COMP%]{margin-top:7px;margin-left:3px;display:inline-block}.pagination-controllers[_ngcontent-%COMP%]{float:right}.pagination-controllers[_ngcontent-%COMP%]   input[_ngcontent-%COMP%]{min-width:60px;text-align:right}.pagination-limit[_ngcontent-%COMP%]{margin-right:25px;display:inline-table;width:150px}.pagination-pages[_ngcontent-%COMP%]{display:inline-block}.pagination-page[_ngcontent-%COMP%]{width:110px;display:inline-table}.pagination-page[_ngcontent-%COMP%]   .input-group-addon[_ngcontent-%COMP%]{display:inline}.pagination-box[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]{outline:0!important}.pagination-firstpage[_ngcontent-%COMP%], .pagination-lastpage[_ngcontent-%COMP%], .pagination-nextpage[_ngcontent-%COMP%], .pagination-prevpage[_ngcontent-%COMP%]{vertical-align:top}.pagination-reload[_ngcontent-%COMP%]{color:gray;font-size:12px}"] });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(DataTablePagination, [{
        type: Component,
        args: [{
                selector: 'data-table-pagination',
                templateUrl: './pagination.html',
                styleUrls: ['./pagination.css']
            }]
    }], function () { return [{ type: i1.DataTable, decorators: [{
                type: Inject,
                args: [forwardRef(() => DataTable)]
            }] }]; }, null); })();
