// ngx-toast-notifications/esm5/toast-notifications.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: toast-notifications.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TOAST_NOTIFICATIONS_CONFIG } from './toast-notifications.config';
import { ToastContainerComponent } from './toast-container/toast-container.component';
import { BasicToastContentComponent } from './toast-content/basic-toast-content.component';
import { ToastContentDirective } from './toast-content/toast-content.directive';
import { Toaster } from './toaster';
import { ToastContainerService } from './toast-container.service';
var ɵ0 = {};
var ToastNotificationsModule = /** @class */ (function () {
    function ToastNotificationsModule(parentModule) {
        if (parentModule) {
            throw new Error('ToastNotificationsModule is already loaded. Import it in the root module only');
        }
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    ToastNotificationsModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: ToastNotificationsModule,
            providers: [
                { provide: TOAST_NOTIFICATIONS_CONFIG, useValue: config },
            ]
        };
    };
    ToastNotificationsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                    ],
                    declarations: [
                        ToastContainerComponent,
                        BasicToastContentComponent,
                        ToastContentDirective,
                    ],
                    entryComponents: [
                        ToastContainerComponent,
                        BasicToastContentComponent,
                    ],
                    providers: [
                        Toaster,
                        ToastContainerService,
                        { provide: TOAST_NOTIFICATIONS_CONFIG, useValue: ɵ0 },
                    ],
                },] }
    ];
    /** @nocollapse */
    ToastNotificationsModule.ctorParameters = function () { return [
        { type: ToastNotificationsModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
    ]; };
    return ToastNotificationsModule;
}());
export { ToastNotificationsModule };
export { ɵ0 };
