// ngx-toast-notifications/esm5/toaster.js
/**
 * @fileoverview added by tsickle
 * Generated from: toaster.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, Type } from '@angular/core';
import { ToastContainerService } from './toast-container.service';
import { TOAST_NOTIFICATIONS_CONFIG } from './toast-notifications.config';
import { BasicToastContentComponent } from './toast-content/basic-toast-content.component';
/** @type {?} */
var DEFAULT_CONFIG = {
    autoClose: true,
    duration: 8000,
    type: 'light',
    position: 'bottom-right',
    component: BasicToastContentComponent,
};
var Toaster = /** @class */ (function () {
    function Toaster(_config, _containerService) {
        this._config = _config;
        this._containerService = _containerService;
    }
    /**
     * @param {?} config
     * @param {?=} componentConfig
     * @return {?}
     */
    Toaster.prototype.open = /**
     * @param {?} config
     * @param {?=} componentConfig
     * @return {?}
     */
    function (config, componentConfig) {
        if (typeof config === 'string') {
            config = tslib_1.__assign({ text: (/** @type {?} */ (config)) }, componentConfig);
        }
        if (config instanceof Type) {
            config = tslib_1.__assign({}, componentConfig, { component: (/** @type {?} */ (config)) });
        }
        config = tslib_1.__assign({}, DEFAULT_CONFIG, this._config, config);
        return this._containerService.ref.instance.add(config);
    };
    Toaster.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    Toaster.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TOAST_NOTIFICATIONS_CONFIG,] }] },
        { type: ToastContainerService }
    ]; };
    return Toaster;
}());
export { Toaster };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Toaster.prototype._config;
    /**
     * @type {?}
     * @private
     */
    Toaster.prototype._containerService;
}
