// angular-bootstrap-data-table/esm5/lib/components/header.component.js
import { Component, Inject, forwardRef } from '@angular/core';
import { DataTable } from './table.component';
import * as i0 from "@angular/core";
import * as i1 from "./table.component";
import * as i2 from "@angular/common";
import * as i3 from "@angular/forms";
function DataTableHeader_div_8_div_1_Template(rf, ctx) { if (rf & 1) {
    var _r6 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 12);
    i0.ɵɵelementStart(1, "label");
    i0.ɵɵelementStart(2, "input", 13);
    i0.ɵɵlistener("ngModelChange", function DataTableHeader_div_8_div_1_Template_input_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r6); var ctx_r5 = i0.ɵɵnextContext(2); return ctx_r5.dataTable.expandColumnVisible = $event; });
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "span");
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngModel", ctx_r1.dataTable.expandColumnVisible);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r1.dataTable.translations.expandColumn);
} }
function DataTableHeader_div_8_div_2_Template(rf, ctx) { if (rf & 1) {
    var _r8 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 12);
    i0.ɵɵelementStart(1, "label");
    i0.ɵɵelementStart(2, "input", 13);
    i0.ɵɵlistener("ngModelChange", function DataTableHeader_div_8_div_2_Template_input_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r8); var ctx_r7 = i0.ɵɵnextContext(2); return ctx_r7.dataTable.indexColumnVisible = $event; });
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "span");
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngModel", ctx_r2.dataTable.indexColumnVisible);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r2.dataTable.translations.indexColumn);
} }
function DataTableHeader_div_8_div_3_Template(rf, ctx) { if (rf & 1) {
    var _r10 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 12);
    i0.ɵɵelementStart(1, "label");
    i0.ɵɵelementStart(2, "input", 13);
    i0.ɵɵlistener("ngModelChange", function DataTableHeader_div_8_div_3_Template_input_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r10); var ctx_r9 = i0.ɵɵnextContext(2); return ctx_r9.dataTable.selectColumnVisible = $event; });
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "span");
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngModel", ctx_r3.dataTable.selectColumnVisible);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r3.dataTable.translations.selectColumn);
} }
function DataTableHeader_div_8_div_4_Template(rf, ctx) { if (rf & 1) {
    var _r13 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 14);
    i0.ɵɵelementStart(1, "label");
    i0.ɵɵelementStart(2, "input", 13);
    i0.ɵɵlistener("ngModelChange", function DataTableHeader_div_8_div_4_Template_input_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r13); var column_r11 = ctx.$implicit; return column_r11.visible = $event; });
    i0.ɵɵelementEnd();
    i0.ɵɵelement(3, "span", 15);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    var column_r11 = ctx.$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngModel", column_r11.visible);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("textContent", column_r11.header);
} }
function DataTableHeader_div_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 9);
    i0.ɵɵtemplate(1, DataTableHeader_div_8_div_1_Template, 5, 2, "div", 10);
    i0.ɵɵtemplate(2, DataTableHeader_div_8_div_2_Template, 5, 2, "div", 10);
    i0.ɵɵtemplate(3, DataTableHeader_div_8_div_3_Template, 5, 2, "div", 10);
    i0.ɵɵtemplate(4, DataTableHeader_div_8_div_4_Template, 4, 2, "div", 11);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r0.dataTable.expandableRows);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r0.dataTable.indexColumn);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r0.dataTable.selectColumn);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngForOf", ctx_r0.dataTable.columns);
} }
var DataTableHeader = /** @class */ (function () {
    function DataTableHeader(dataTable) {
        this.dataTable = dataTable;
        this.columnSelectorOpen = false;
    }
    DataTableHeader.prototype._closeSelector = function () {
        this.columnSelectorOpen = false;
    };
    DataTableHeader.ɵfac = function DataTableHeader_Factory(t) { return new (t || DataTableHeader)(i0.ɵɵdirectiveInject(forwardRef(function () { return DataTable; }))); };
    DataTableHeader.ɵcmp = i0.ɵɵdefineComponent({ type: DataTableHeader, selectors: [["data-table-header"]], hostBindings: function DataTableHeader_HostBindings(rf, ctx) { if (rf & 1) {
            i0.ɵɵlistener("click", function DataTableHeader_click_HostBindingHandler() { return ctx._closeSelector(); }, false, i0.ɵɵresolveDocument);
        } }, decls: 9, vars: 4, consts: [[1, "data-table-header"], [1, "title", 3, "textContent"], [1, "button-panel"], ["type", "button", 1, "btn", "btn-default", "btn-sm", "refresh-button", 3, "click"], [1, "fa", "fa-refresh"], ["type", "button", 1, "btn", "btn-default", "btn-sm", "column-selector-button", 3, "click"], [1, "fa", "fa-list"], [1, "column-selector-wrapper", 3, "click"], ["class", "column-selector-box card", 4, "ngIf"], [1, "column-selector-box", "card"], ["class", "column-selector-fixed-column checkbox", 4, "ngIf"], ["class", "column-selector-column checkbox", 4, "ngFor", "ngForOf"], [1, "column-selector-fixed-column", "checkbox"], ["type", "checkbox", 3, "ngModel", "ngModelChange"], [1, "column-selector-column", "checkbox"], [3, "textContent"]], template: function DataTableHeader_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵelement(1, "h4", 1);
            i0.ɵɵelementStart(2, "div", 2);
            i0.ɵɵelementStart(3, "button", 3);
            i0.ɵɵlistener("click", function DataTableHeader_Template_button_click_3_listener() { return ctx.dataTable.reloadItems(); });
            i0.ɵɵelement(4, "i", 4);
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(5, "button", 5);
            i0.ɵɵlistener("click", function DataTableHeader_Template_button_click_5_listener($event) { ctx.columnSelectorOpen = !ctx.columnSelectorOpen; return $event.stopPropagation(); });
            i0.ɵɵelement(6, "i", 6);
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(7, "div", 7);
            i0.ɵɵlistener("click", function DataTableHeader_Template_div_click_7_listener($event) { return $event.stopPropagation(); });
            i0.ɵɵtemplate(8, DataTableHeader_div_8_Template, 5, 4, "div", 8);
            i0.ɵɵelementEnd();
            i0.ɵɵelementEnd();
            i0.ɵɵelementEnd();
        } if (rf & 2) {
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("textContent", ctx.dataTable.headerTitle);
            i0.ɵɵadvance(4);
            i0.ɵɵclassProp("active", ctx.columnSelectorOpen);
            i0.ɵɵadvance(3);
            i0.ɵɵproperty("ngIf", ctx.columnSelectorOpen);
        } }, directives: [i2.NgIf, i2.NgForOf, i3.CheckboxControlValueAccessor, i3.NgControlStatus, i3.NgModel], styles: [".data-table-header[_ngcontent-%COMP%]{min-height:25px;margin-bottom:10px}.title[_ngcontent-%COMP%]{display:inline-block;margin:5px 0 0 5px}.button-panel[_ngcontent-%COMP%]{float:right}.button-panel[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]{outline:0!important}.column-selector-wrapper[_ngcontent-%COMP%]{position:relative}.column-selector-box[_ngcontent-%COMP%]{box-shadow:0 0 10px #d3d3d3;width:150px;padding:10px;position:absolute;right:0;top:1px;z-index:1060}.column-selector-box[_ngcontent-%COMP%]   .checkbox[_ngcontent-%COMP%]{margin-bottom:4px}.column-selector-fixed-column[_ngcontent-%COMP%]{font-style:italic}"] });
    return DataTableHeader;
}());
export { DataTableHeader };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(DataTableHeader, [{
        type: Component,
        args: [{
                selector: 'data-table-header',
                templateUrl: './header.html',
                styleUrls: ['./header.css'],
                host: {
                    '(document:click)': '_closeSelector()'
                }
            }]
    }], function () { return [{ type: i1.DataTable, decorators: [{
                type: Inject,
                args: [forwardRef(function () { return DataTable; })]
            }] }]; }, null); })();
