// angular-bootstrap-data-table/esm5/lib/components/row.component.js
import { Component, Input, Inject, forwardRef, Output, EventEmitter } from '@angular/core';
import { DataTable } from './table.component';
import * as i0 from "@angular/core";
import * as i1 from "./table.component";
import * as i2 from "../utils/hide";
import * as i3 from "@angular/forms";
import * as i4 from "@angular/common";
var _c0 = ["dataTableRow", ""];
function DataTableRow_td_7_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 12);
} if (rf & 2) {
    var column_r2 = i0.ɵɵnextContext().$implicit;
    var ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵproperty("textContent", ctx_r3.item[column_r2.property]);
} }
var _c1 = function (a0, a1, a2) { return { column: a0, row: a1, item: a2 }; };
function DataTableRow_td_7_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 13);
} if (rf & 2) {
    var column_r2 = i0.ɵɵnextContext().$implicit;
    var ctx_r4 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngTemplateOutlet", column_r2.cellTemplate)("ngTemplateOutletContext", i0.ɵɵpureFunction3(2, _c1, column_r2, ctx_r4._this, ctx_r4.item));
} }
function DataTableRow_td_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "td", 9);
    i0.ɵɵtemplate(1, DataTableRow_td_7_div_1_Template, 1, 1, "div", 10);
    i0.ɵɵtemplate(2, DataTableRow_td_7_div_2_Template, 1, 6, "div", 11);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    var column_r2 = ctx.$implicit;
    var ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("background-color", column_r2.getCellColor(ctx_r0._this, ctx_r0.index));
    i0.ɵɵproperty("hide", !column_r2.visible)("ngClass", column_r2.styleClassObject);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !column_r2.cellTemplate);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", column_r2.cellTemplate);
} }
var _c2 = function (a0, a1) { return { row: a0, item: a1 }; };
function DataTableRow_tr_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "tr", 14);
    i0.ɵɵelementStart(1, "td");
    i0.ɵɵelement(2, "div", 13);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("hide", !ctx_r1.expanded);
    i0.ɵɵadvance(1);
    i0.ɵɵattribute("colspan", ctx_r1.dataTable.columnCount);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r1.dataTable.expandTemplate)("ngTemplateOutletContext", i0.ɵɵpureFunction2(4, _c2, ctx_r1._this, ctx_r1.item));
} }
var DataTableRow = /** @class */ (function () {
    function DataTableRow(dataTable) {
        this.dataTable = dataTable;
        this.selectedChange = new EventEmitter();
        this._this = this; // FIXME is there no template keyword for this in angular 2?
    }
    Object.defineProperty(DataTableRow.prototype, "selected", {
        get: function () {
            return this._selected;
        },
        set: function (selected) {
            this._selected = selected;
            this.selectedChange.emit(selected);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTableRow.prototype, "displayIndex", {
        // other:
        get: function () {
            if (this.dataTable.pagination) {
                return this.dataTable.displayParams.offset + this.index + 1;
            }
            else {
                return this.index + 1;
            }
        },
        enumerable: true,
        configurable: true
    });
    DataTableRow.prototype.getTooltip = function () {
        if (this.dataTable.rowTooltip) {
            return this.dataTable.rowTooltip(this.item, this, this.index);
        }
        return '';
    };
    DataTableRow.prototype.ngOnDestroy = function () {
        this.selected = false;
    };
    DataTableRow.ɵfac = function DataTableRow_Factory(t) { return new (t || DataTableRow)(i0.ɵɵdirectiveInject(forwardRef(function () { return DataTable; }))); };
    DataTableRow.ɵcmp = i0.ɵɵdefineComponent({ type: DataTableRow, selectors: [["", "dataTableRow", ""]], inputs: { item: "item", index: "index" }, outputs: { selectedChange: "selectedChange" }, attrs: _c0, decls: 9, vars: 20, consts: [[1, "data-table-row", 3, "title", "dblclick", "click"], [1, "row-expand-button", 3, "hide", "click"], [1, "fa", "fa-arrow-circle-o-right", 3, "hide"], [1, "fa", "fa-arrow-circle-o-down", 3, "hide"], [1, "index-column", 3, "hide", "textContent"], [1, "select-column", 3, "hide"], ["type", "checkbox", 3, "ngModel", "ngModelChange"], ["class", "data-column", 3, "hide", "ngClass", "background-color", 4, "ngFor", "ngForOf"], ["class", "row-expansion", 3, "hide", 4, "ngIf"], [1, "data-column", 3, "hide", "ngClass"], [3, "textContent", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext", 4, "ngIf"], [3, "textContent"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "row-expansion", 3, "hide"]], template: function DataTableRow_Template(rf, ctx) { if (rf & 1) {
            i0.ɵɵelementStart(0, "tr", 0);
            i0.ɵɵlistener("dblclick", function DataTableRow_Template_tr_dblclick_0_listener($event) { return ctx.dataTable.rowDoubleClicked(ctx._this, $event); })("click", function DataTableRow_Template_tr_click_0_listener($event) { return ctx.dataTable.rowClicked(ctx._this, $event); });
            i0.ɵɵelementStart(1, "td", 1);
            i0.ɵɵlistener("click", function DataTableRow_Template_td_click_1_listener($event) { ctx.expanded = !ctx.expanded; return $event.stopPropagation(); });
            i0.ɵɵelement(2, "span", 2);
            i0.ɵɵelement(3, "span", 3);
            i0.ɵɵelementEnd();
            i0.ɵɵelement(4, "td", 4);
            i0.ɵɵelementStart(5, "td", 5);
            i0.ɵɵelementStart(6, "input", 6);
            i0.ɵɵlistener("ngModelChange", function DataTableRow_Template_input_ngModelChange_6_listener($event) { return ctx.selected = $event; });
            i0.ɵɵelementEnd();
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(7, DataTableRow_td_7_Template, 3, 6, "td", 7);
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(8, DataTableRow_tr_8_Template, 3, 7, "tr", 8);
        } if (rf & 2) {
            i0.ɵɵstyleProp("background-color", ctx.dataTable.getRowColor(ctx.item, ctx.index, ctx._this));
            i0.ɵɵclassProp("row-odd", ctx.index % 2 === 0)("row-even", ctx.index % 2 === 1)("selected", ctx.selected)("clickable", ctx.dataTable.selectOnRowClick);
            i0.ɵɵproperty("title", ctx.getTooltip());
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("hide", !ctx.dataTable.expandColumnVisible);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("hide", ctx.expanded);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("hide", !ctx.expanded);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("hide", !ctx.dataTable.indexColumnVisible)("textContent", ctx.displayIndex);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("hide", !ctx.dataTable.selectColumnVisible);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngModel", ctx.selected);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngForOf", ctx.dataTable.columns);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx.dataTable.expandableRows);
        } }, directives: [i2.Hide, i3.CheckboxControlValueAccessor, i3.NgControlStatus, i3.NgModel, i4.NgForOf, i4.NgIf, i4.NgClass, i4.NgTemplateOutlet], styles: [".select-column[_ngcontent-%COMP%]{text-align:center}.row-expand-button[_ngcontent-%COMP%]{cursor:pointer;text-align:center}.clickable[_ngcontent-%COMP%]{cursor:pointer}"] });
    return DataTableRow;
}());
export { DataTableRow };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(DataTableRow, [{
        type: Component,
        args: [{
                selector: '[dataTableRow]',
                templateUrl: './row.html',
                styleUrls: ['./row.css']
            }]
    }], function () { return [{ type: i1.DataTable, decorators: [{
                type: Inject,
                args: [forwardRef(function () { return DataTable; })]
            }] }]; }, { item: [{
            type: Input
        }], index: [{
            type: Input
        }], selectedChange: [{
            type: Output
        }] }); })();
