// ng-img-magnifier/bundles/ng-img-magnifier.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('ng-img-magnifier', ['exports', '@angular/core', '@angular/common', '@angular/platform-browser'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ng-img-magnifier'] = {}, global.ng.core, global.ng.common, global.ng.platformBrowser));
}(this, (function (exports, core, common, platformBrowser) { 'use strict';

    var NgImgMagnifierComponent = /** @class */ (function () {
        function NgImgMagnifierComponent() {
            this.thumbImage = '';
            this.fullImage = '';
            this.top = '2';
            this.right = '1';
            this.lensWidth = '150';
            this.lensHeight = '150';
            this.resultWidth = '35';
            this.resultHeight = '60';
            this.imgWidth = '300';
            this.imgHeight = '300';
        }
        NgImgMagnifierComponent.prototype.ngOnInit = function () {
        };
        NgImgMagnifierComponent.prototype.imageZoom = function () {
            this.showLensDiv();
            var img, lens, result, cx, cy;
            result = this.resultID.nativeElement;
            result.style.top = this.top + '%';
            result.style.right = this.right + '%';
            result.style.width = this.resultWidth + '%';
            result.style.height = this.resultHeight + '%';
            this.showZoom = true;
            img = this.imgID.nativeElement;
            lens = this.zoomDiv.nativeElement;
            if (img.width > (img.height * 2)) {
                lens.style.width = (img.width / 3) + "px";
                lens.style.height = (img.height / 1) + "px";
            }
            else if (img.height > (img.width * 2)) {
                lens.style.width = (img.width / 1) + "px";
                lens.style.height = (img.height / 3) + "px";
            }
            else {
                lens.style.width = this.lensWidth + "px";
                lens.style.height = this.lensHeight + "px";
            }
            lens.style.opacity = 1;
            cx = result.offsetWidth / lens.offsetWidth;
            cy = result.offsetHeight / lens.offsetHeight;
            result.style.backgroundImage = "url('" + this.fullImage + "')";
            result.style.backgroundSize = (img.width * cx) + "px " + (img.height * cy) + "px";
            lens.addEventListener("mousemove", moveLens);
            img.addEventListener("mousemove", moveLens);
            lens.addEventListener("touchmove", moveLens);
            img.addEventListener("touchmove", moveLens);
            function moveLens(e) {
                var pos, x, y;
                e.preventDefault();
                pos = getPointerPosition(e);
                x = pos.x - (lens.offsetWidth / 2);
                y = pos.y - (lens.offsetHeight / 2);
                if (x > img.width - lens.offsetWidth) {
                    x = img.width - lens.offsetWidth;
                }
                if (x < 0) {
                    x = 0;
                }
                if (y > img.height - lens.offsetHeight) {
                    y = img.height - lens.offsetHeight;
                }
                if (y < 0) {
                    y = 0;
                }
                lens.style.left = (x + 5) + "px";
                lens.style.top = (y + 5) + "px";
                result.style.backgroundPosition = "-" + (x * cx) + "px -" + (y * cy) + "px";
            }
            function getPointerPosition(e) {
                var a, x = 0, y = 0;
                e = e || window.event;
                a = img.getBoundingClientRect();
                x = e.pageX - a.left;
                y = e.pageY - a.top;
                x = x - window.pageXOffset;
                y = y - window.pageYOffset;
                return { x: x, y: y };
            }
        };
        NgImgMagnifierComponent.prototype.showZoomWin = function () {
            this.showZoom = true;
        };
        NgImgMagnifierComponent.prototype.hideZoomWin = function () {
            this.showZoom = false;
            this.zoomDiv.nativeElement.style.opacity = 0;
            this.zoomDiv.nativeElement.style.display = 'none';
        };
        NgImgMagnifierComponent.prototype.showLensDiv = function () {
            this.zoomDiv.nativeElement.style.display = 'block';
        };
        return NgImgMagnifierComponent;
    }());
    NgImgMagnifierComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'ng-img-magnifier',
                    template: "\n    <div class=\"img-zoom-container \">\n      <div class=\"outerImg \" (mouseenter)=\"showZoomWin() \" (mouseleave)=\"hideZoomWin() \">\n          <img #myimage id=\"myimage\" class=\"mainImg\" src={{thumbImage}} width={{imgWidth}} style=\"vertical-align: middle !important;\" height={{imgHeight}} (mousemove)=\"imageZoom();\">\n          <div #zoomDiv id=\"zoomDiv \" class=\"img-zoom-lens \"></div>\n      </div>\n    </div>\n    <div *ngIf=\"showZoom \">\n        <div #myresult id=\"myresult \" class=\"img-zoom-result \"></div>\n    </div>",
                    styles: ["html,\n    body {\n        height: 100%;\n    }",
                        "body {\n        margin: 0;\n    }",
                        "* {\n        box-sizing: border-box;\n    }",
                        ".img-zoom-container {\n        position: relative;\n    }",
                        ".img-zoom-lens {\n        position: absolute;\n        border: 1px solid #d4d4d4;\n        // background-image: url('./assets/zoom-image.png');\n        background-color: rgb(255, 255, 255, 0.2);\n        background-repeat: repeat;\n        background-size: contain;\n        opacity: 0;\n        cursor: zoom-in !important;\n        display: none;\n    }",
                        ".img-zoom-result {\n        background-color: none;\n        background-repeat: no-repeat;\n        transition: 0.3s;\n        border-radius: 5px;\n        position: absolute;\n    }",
                        "@media screen and (max-width: 800px) {\n        .img-zoom-result {\n            display: none;\n        }\n    }",
                        ".show {\n        opacity: 500;\n    }",
                        ".hide {\n        opacity: 0;\n    }",
                        ".mainImg {\n        vertical-align: middle !important;\n    }",
                        ".outerImg {\n        width: 310px;\n        height: auto;\n        padding-left: 5px;\n        padding-right: 5px;\n        padding-top: 5px;\n        padding-bottom: 5px;\n    }"]
                },] }
    ];
    NgImgMagnifierComponent.ctorParameters = function () { return []; };
    NgImgMagnifierComponent.propDecorators = {
        thumbImage: [{ type: core.Input }],
        fullImage: [{ type: core.Input }],
        top: [{ type: core.Input }],
        right: [{ type: core.Input }],
        lensWidth: [{ type: core.Input }],
        lensHeight: [{ type: core.Input }],
        resultWidth: [{ type: core.Input }],
        resultHeight: [{ type: core.Input }],
        imgWidth: [{ type: core.Input }],
        imgHeight: [{ type: core.Input }],
        resultID: [{ type: core.ViewChild, args: ['myresult',] }],
        imgID: [{ type: core.ViewChild, args: ['myimage',] }],
        zoomDiv: [{ type: core.ViewChild, args: ['zoomDiv',] }]
    };

    var ImgMagnifier = /** @class */ (function () {
        function ImgMagnifier() {
        }
        return ImgMagnifier;
    }());
    ImgMagnifier.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [
                        NgImgMagnifierComponent,
                    ],
                    imports: [
                        common.CommonModule,
                        platformBrowser.BrowserModule
                    ],
                    exports: [
                        NgImgMagnifierComponent
                    ]
                },] }
    ];

    /*
     * Public API Surface of ng-img-magnifier
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ImgMagnifier = ImgMagnifier;
    exports.NgImgMagnifierComponent = NgImgMagnifierComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
