// ngx-toast-notifications/esm2015/toast.js
/**
 * @fileoverview added by tsickle
 * Generated from: toast.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
export class Toast {
    /**
     * @param {?} config
     * @param {?} closeFunction
     */
    constructor(config, closeFunction) {
        this._onClose = new Subject();
        this.autoClose = config.autoClose;
        this.duration = config.duration > 0 ? config.duration : 0;
        this.text = config.text;
        this.caption = config.caption;
        this.type = config.type;
        this.component = config.component;
        this._closeFunction = closeFunction;
        this._setTimeout();
    }
    /**
     * @return {?}
     */
    get onClose() {
        return this._onClose.asObservable();
    }
    /**
     * @param {?=} result
     * @return {?}
     */
    close(result) {
        if (!this._onClose.closed) {
            this._onClose.next(result);
            this._onClose.complete();
        }
        this._closeFunction(this);
        this._clearTimeout();
    }
    /**
     * @private
     * @return {?}
     */
    _setTimeout() {
        if (this.autoClose && this.duration > 0) {
            this._timeoutId = setTimeout((/**
             * @return {?}
             */
            () => this.close()), this.duration);
        }
    }
    /**
     * @private
     * @return {?}
     */
    _clearTimeout() {
        if (this._timeoutId) {
            clearTimeout(this._timeoutId);
        }
    }
}
if (false) {
    /** @type {?} */
    Toast.prototype.autoClose;
    /** @type {?} */
    Toast.prototype.duration;
    /** @type {?} */
    Toast.prototype.text;
    /** @type {?} */
    Toast.prototype.caption;
    /** @type {?} */
    Toast.prototype.type;
    /** @type {?} */
    Toast.prototype.component;
    /**
     * @type {?}
     * @private
     */
    Toast.prototype._closeFunction;
    /**
     * @type {?}
     * @private
     */
    Toast.prototype._onClose;
    /**
     * @type {?}
     * @private
     */
    Toast.prototype._timeoutId;
}
